module/zfs/dmu_send.c

```diff
@@ -1130,15 +1138,24 @@ dmu_send_impl(void *tag, dsl_pool_t *dp, dsl_dataset_t *to_ds,
 			err = dsl_crypto_populate_key_nvlist(to_ds, &keynvl);
 			if (err != 0) {
 				fnvlist_free(nvl);
+				fnvlist_free(holds_nvl);
 				goto out;
 			}
 
 			fnvlist_add_nvlist(nvl, "crypt_keydata", keynvl);
 		}
 
+		if (featureflags & DMU_BACKUP_FEATURE_HOLDS) {
+			char *exclude = ".send";
```

Comment:
I guess this is needed to skip holds created by libzfs as noted here:

https://github.com/zfsonlinux/zfs/blob/d1ed1be3cde1be1d4016c9725e9ca0920a3350af/cmd/zfs/zfs_main.c#L5615

Maybe we should make this explicit with a comment here.

Also, given the updated man page states

> Generate a stream package that includes any snapshot holds (created with the `zfs hold` command)

we should skip all "libzfs" holds (everything starting with a [dot], not ".send") since they cannot be created with `zfs hold`.

Reply:
Also this doesn't seem to work when receiving a stream created with `zfs send -R`:

```
root@linux:~# TMPDIR='/var/tmp'
root@linux:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@linux:~# zpool destroy $POOLNAME
root@linux:~# rm -f $TMPDIR/disk.dat
root@linux:~# truncate -s 128m $TMPDIR/disk.dat
root@linux:~# zpool create $POOLNAME $TMPDIR/disk.dat
root@linux:~# 
root@linux:~# zfs create $POOLNAME/send
root@linux:~# zfs snap $POOLNAME/send@s1
root@linux:~# zfs send -h $POOLNAME/send@s1 | zfs recv $POOLNAME/recv1
root@linux:~# zfs send -Rh $POOLNAME/send@s1 | zfs recv $POOLNAME/recv2
root@linux:~# 
root@linux:~# zfs list -H -t snap -o name | xargs -L1 zfs holds -H
testpool/recv2@s1	.send-11664-1	Sat Aug 25 12:26 2018
root@linux:~# 
```

Reply:
Hi @loli10K are you sure you are running the latest patch?  This may have been an issue in the earlier version.  I can't reproduce it now.

Reply:
I can't reproduce either on a clean builder, i was probably running some old bits on that builder. Sorry for the confusion.